cluster/cluster.go

```diff
@@ -15,4 +38,85 @@ func Init(name, version string, primary bool, started time.Time) {
 		primaryChange: time.Now(),
 		stateChange:   time.Now(),
 	}
+	Mode = m
+	go probePeers(interval)
+}
+
+func Stop() {
+	close(shutdown)
+}
+
+func AddPeer(remoteAddr *url.URL) {
+	mu.Lock()
+	peer := &Node{
+		remoteAddr:  remoteAddr,
+		stateChange: time.Now(),
+	}
+	peers = append(peers, peer)
+	go peer.Probe()
+	mu.Unlock()
+}
+
+func GetPeers() []*Node {
+	mu.Lock()
+	p := make([]*Node, len(peers))
+	copy(p, peers)
+	mu.Unlock()
+	return p
+}
+
+func probePeers(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for {
+		select {
+		case <-shutdown:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			mu.Lock()
+			for _, peer := range peers {
+				go peer.Probe()
+			}
+			mu.Unlock()
+		}
+	}
+}
+
+// return the list of peers to broadcast requests to
+// Only 1 peer per partition is returned
+func PeersForQuery() []*Node {
+	// If we are running in single mode, just return thisNode
+	if Mode == ModeSingle {
+		return []*Node{ThisNode}
+	}
+
+	peersMap := make(map[int32][]*Node)
+	for _, part := range ThisNode.GetPartitions() {
+		peersMap[part] = []*Node{ThisNode}
+	}
+	mu.Lock()
+	for _, peer := range peers {
+		if !peer.IsReady() {
+			continue
+		}
+		for _, part := range peer.GetPartitions() {
+			peersMap[part] = append(peersMap[part], peer)
+		}
+	}
+	mu.Unlock()
+	selectedPeers := make(map[string]*Node)
+	answer := make([]*Node, 0)
+	// we only need to query 1 node for each partition.
+	// we are just using the first node for each partition, which will be
+	// ThisNode if ThisNode owns the partition.  Later we can add more
+	// complex scheduling logic to distribute queries across the cluster
```

Comment:
another problem with my approach is assigning primaries becomes hard.
with my proposal above, we could make A and D primaries, so all partitions are getting saved, and only one time.
but if A crashes and we want to establish a new primary for part 0 and 1, we would have to select B and C, but those would also save data that D is already saving. we could de-promote D and promote B and C, but then what if either of those fails. (this reminds me a lot of raid levels btw).

I think to implement my approach, we would need to support primary status on a per-partition level instead of per-node. so maybe for now we should stick to your way.